content/blog/2019-12-18-polynomial/index.md

```diff
@@ -0,0 +1,155 @@
+performance FFT implementation, like [FFTW](http://www.fftw.org/) and
+[Spiral](https://www.spiral.net/index.html). In recent years, Image
+processing and machine learning communities also propose several framework to
+generate high performance code, such as on the
+[Halide](https://github.com/halide/Halide), [Loopy](https://github.com/inducer/loopy) and
+[TVM](https://github.com/apache/incubator-tvm). 
+
+## Method 
+
+### Schoolbook Multiplication
+
+This is the straight forward basic method.
+
+### Karatsuba
+We break the polynomial $f(x)$ and $g(x)$ to high and low two part.
+
+
+  $$
+  f(x) = a_1(x) t + a_0(x)
+  $$
+  $$
+  g(x) = b_1(x) t + b_0(x)
+  $$
+  , where $t = x^{\frac{n}{2}}$. 
+  Then the multiplication can be calculated as follows:
+
+  $$
+  f(x)g(x) = c_2(x) t^2 + c_1 t + c_0(x)
+  $$
+
+  where 
+  $$ c_2(x) = a_1(x)b_1(x) $$
+  $$ c_0(x) = a_0(x)b_0(x) $$
+
+  $$
+  c_1(x) = (a_1(x) + a_0(x)) (b_1(x) +b_0(x)) -c_0(x)-c_2(x)
+  $$
+
+  It uses 3 multiplications while the schoolbook method would use 4.
+  The idea can be further generalize to k-way split, and it is called Toom-k.
+
+### Toom-k
+
+Lets take Toom-3 as an example. First we divide the polynomial $f(x)$ and $g(x)$ into 3 parts.
+
+$$
+f(x) = a_2(x) t^2 + a_1(x) t + a_0(x)
+$$
+
+$$
+g(x) = b_2(x) t^2 + b_1(x) t + b_0(x)
+$$
+
+where $t = x^{\frac{n}{3}}$. 
+
+
+Now, we substitute t with values: $0$, $1$, $-1$, $2$, and $\inf$.
+and we let $f(x)g(x) = c_4(x)t^4 + c_3(x)t^3 + c_2(x) t^2 + c_1(x) t + c_0(x)$
+We get the following table:
+
+|  $t=$    |                                                      |                          |
+|---------|------------------------------------------------------|------------------------------------------------|
+|$ 0 $  |$a_0(x)b_0(x)$                                        |$=c_0(x)$                                        |
+|$ 1 $  |$(a_2(x)+a_1(x)+a_0(x)) (b_2(x)+b_1(x)+b_0(x))$       |$=c_4(x) +c_3(x)+c_2(x)+c_1(x)+c_0(x)  $         |
+|$ -1$  |$(a_2(x)-a_1(x)+a_0(x)) (b_2(x)-b_1(x)+b_0(x))$       |$=c_4(x) -c_3(x)+c_2(x) -c_1(x)+c_0(x)  $        |
+|$ 2 $  |$(4a_2(x)+2a_1(x)+a_0(x)) (4b_2(x)+2b_1(x)+b_0(x))$   |$=16c_4(x) +8c_3(x)+4c_2(x)+2c_1(x)+c_0(x)  $    |
+|$\inf$ |$a_2(x)b_2(x)$                                        |$=c_4(x)$                                        |
+
+The second column is the 5 multiplications we perform, and then we can see that
+we can solve an linear systems to get $c_4,c_3,c_2,c_1,c_0$ and mainly only use
+addition, subtraction and constant multiplication. In this case, the number of
+multiplication is 5, instead of 9 in the case of schoolbook method. We here
+evaluate at 5 points, but this method can be generalized to $k$-way split evaluate
+on $2k-1$ points.
+
+### FFT
+
+As we are dealing with polynomial with small to medium degree, we do not
+consider the FFT method here. Yet it will be interesting to combine this with
```

Comment:
I don't know what "the FFT method" is.

Reply:
Remove this part as FFT is not considered in this project.